src/charm.py

```diff
@@ -83,119 +129,78 @@ def _training_operator_layer(self) -> Layer:
+            self.container.add_layer(self._container_name, new_layer, combine=True)
+            try:
+                self.logger.info("Pebble plan updated with new configuration, replaning")
+                self.container.replan()
+            except ChangeError:
+                raise ErrorWithStatus("Failed to replan", BlockedStatus)
+
+    def main(self, _) -> None:
+        """Perform all required actions the Charm."""
         try:
-            self._create_resource(resource_type="auth", context=self._context)
-        except ApiError as e:
-            if e.status.reason == "AlreadyExists":
-                logging.info(
-                    f"{e.status.details.name} auth resource already present. It will be reused."
-                )
-            else:
-                raise
-
-    def _on_install(self, event):
-        """Event handler for InstallEvent."""
+            self._check_container_connection()
+            self._check_leader()
+            self._deploy_k8s_resources()
+            self._update_layer()
+        except ErrorWithStatus as error:
+            self.model.unit.status = error.status
+            return
 
-        # Update Pebble configuration layer if it has changed
-        self._update_layer()
+        self.model.unit.status = ActiveStatus()
 
-        # Patch/create Kubernetes resources
-        try:
-            self.unit.status = MaintenanceStatus("Creating auth resources")
-            self._create_auth_resources()
-            self.unit.status = MaintenanceStatus("Creating CRDs")
-            self._create_crds()
-        except ApiError as e:
-            logging.error(traceback.format_exc())
-            self.unit.status = BlockedStatus(
-                f"Creating/patching resources failed with code {str(e.status.code)}."
+    def _on_pebble_ready(self, _):
+        """Configure started container."""
+        if not self.container.can_connect():
+            raise ErrorWithStatus(
+                f"Container {self._container_name} failed to start", BlockedStatus
             )
-            if e.status.code == 403:
-                logging.error(
-                    "Received Forbidden (403) error when creating auth resources."
-                    "This may be due to the charm lacking permissions to create"
-                    "cluster-scoped resources."
-                    "Charm must be deployed with --trust"
-                )
-                event.defer()
-                return
-        else:
-            self.unit.status = ActiveStatus()
-
-    def _on_config_changed(self, _):
-        """Event handler for ConfigChangedEvent"""
-
-        # Update Pebble configuration layer if it has changed
-        self._update_layer()
-
-        # Patch/create Kubernetes resources
+        self.main(_)
+
+    def _on_install(self, _):
+        """Perform installation only actions."""
+        self._check_container_connection()
+        self._on_pebble_ready(_)
```

Comment:
Looks good. Just a comment: Install event happens before container is started. So, chances are that container is not up at this point. With protection of `_check_container_connection()` we should be good calling Pebble Ready event.

Other solution would be let Pebble Ready event handler to handle items when container is ready and in Install event handler install K8S resources to speed up deployment.
Seldon install handler re-design does this in latest version. In previous version it was following same pattern as here.

Reply:
Sooo leaving it for now ... we can change it later when facing depoyment slowdowns :)